cranelift/codegen/src/isa/zkasm/inst/emit.rs

```diff
@@ -591,14 +591,74 @@ impl MachInstEmit for Inst {
                 match from {
                     AMode::RegOffset(r, ..) => {
                         debug_assert_eq!(r, e0());
+                        // TODO(#43): Implement the conversion using verifiable computations.
+                        put_string(
+                            &format!(
+                                "${{ ({}) / 8 }} => {}\n",
+                                access_reg_with_offset(r, offset),
+                                reg_name(r)
+                            ),
+                            sink,
+                        );
+                        let rem = offset % 8;
+                        let width = op.width() as i64;
+
                         put_string(
                             &format!(
                                 "$ => {} :MLOAD(MEM:{})\n",
                                 reg_name(rd.to_reg()),
-                                access_reg_with_offset(r, offset)
+                                reg_name(r)
                             ),
                             sink,
                         );
+                        // TODO(#34): Implement unaligned and narrow reads using verifiable
+                        // computations.
+                        if rem > 0 {
+                            put_string(
+                                &format!(
+                                    "${{ {} >> {} }} => {}\n",
+                                    reg_name(rd.to_reg()),
+                                    8 * rem,
+                                    reg_name(rd.to_reg())
+                                ),
+                                sink,
+                            );
+                        }
+
+                        // Handle the case when read spans two slots.
+                        if rem + width > 8 {
```

Comment:
Can you really rely on this? What happens if wasm goes like `i32.load offset=0 (i32.const 7)`? I think there's no choice but to check alignment of the final computed (byte) address at runtime and/or unconditionally load two slots.

Reply:
Fair point, I'm not sure whether I can rely on the fact that the address in the load/store register will be word-aligned. So we at least need an assert to check this.

Given that this is a non-trivial amount of work that we might need to redo anyway soon, I'll start with an assert in the code to check that the address is word-aligned. If it is not triggered by the Rust benchmarks that we use, that would be interesting evidence to investigate how exactly these addresses are generated. If it is triggered, we'll need to address it when we have the test/benchmark that exercises it.

Reply:
FWIW for a minimal reproducer something like [`read_unaligned`](https://doc.rust-lang.org/stable/std/ptr/fn.read_unaligned.html) would do. In architectures that require aligned loads the compiler backend would responsibly split the load into aligned loads, but WebAssembly specifically does not impose that sort of requirement, so the backend doesn't need to worry about unaligned addresses at all.

Reply:
It took some effort, but I rewrote the code without the assumption that the address is word-aligned.
I ended up unconditionally reading and writing to two slots, though I think in the future it would not be that hard to optimize the code to skip the second write when possible using a jump.